statsd/worker.go

```diff
@@ -71,9 +73,15 @@ func (w *worker) processMetric(m metric) error {
 }
 
 func (w *worker) shouldSample(rate float64) bool {
+	// rand.NewSource is not thread safe.
+	// TODO: use defer once the lowest Go version we support is 1.14 (defer
+	// has an overhead before that).
+	w.randomLock.Lock()
 	if rate < 1 && w.random.Float64() > rate {
+		w.randomLock.Unlock()
 		return false
 	}
+	w.randomLock.Unlock()
```

Comment:
Just a note: It can be rewrite as
```go
w.randomLock.Lock()
defer w.randomLock.Unlock()
return rate >= 1 || w.random.Float64() <= rate
```

Reply:
it's better to avoid `defer`, see my earlier comment: https://github.com/DataDog/datadog-go/pull/179#discussion_r565463981

Reply:
See comment from olivier about defer overhead